### ISSUE TITLE
Zero-shot Generalization Performance on middlebury

### ISSUE THREAD

**Issue by yyYcoco2** [Insta360-Research-Team/DEFOM-Stereo#13]
请问您是如何计算

![Image](https://github.com/user-attachments/assets/b482791f-83e6-4186-a2a2-63b280e9b435)
这些模型的D1的
例如：IGEV:
@torch.no_grad()
def validate_middlebury(model, iters=32, split='F', mixed_prec=False):
    """ Peform validation using the Middlebury-V3 dataset """
    model.eval()
    aug_params = {}
    val_dataset = datasets.Middlebury(aug_params, split=split)

    out_list, epe_list = [], []
    for val_id in range(len(val_dataset)):
        (imageL_file, _, _), image1, image2, flow_gt, valid_gt = val_dataset[val_id]
        image1 = image1[None].cuda()#F:[1,3,1984,2872] H:[1,3,992,1436] Q:[1,3,496,718]
        image2 = image2[None].cuda()
        print(image1.shape)

        padder = InputPadder(image1.shape, divis_by=32)
        image1, image2 = padder.pad(image1, image2)#F:[1,3,1984,2880] H:[1,3,992,1440]Q:[1,3,512,736]
        print(image1.shape)

        with autocast(enabled=mixed_prec):
            flow_pr = model(image1, image2, iters=iters, test_mode=True)
            print(flow_pr.shape)
        flow_pr = padder.unpad(flow_pr).cpu().squeeze(0) 
        print(f"预测最大视差值: {flow_pr.max().item():.6f}")

        assert flow_pr.shape == flow_gt.shape, (flow_pr.shape, flow_gt.shape)
        epe = torch.sum((flow_pr - flow_gt)**2, dim=0).sqrt()
        epe_flattened = epe.flatten()
        val = flow_gt[0].reshape(-1) < 1000
        out = (epe_flattened > 2.0) #bad 2.0
        image_out = out[val].float().mean().item() 
        image_epe = epe_flattened[val].mean().item()
        logging.info(f"Middlebury Iter {val_id+1} out of {len(val_dataset)}. EPE {round(image_epe,4)} D1 {round(image_out,4)}")
        epe_list.append(image_epe)
        out_list.append(image_out)

    epe_list = np.array(epe_list)
    out_list = np.array(out_list)

    epe = np.mean(epe_list)
    d1 = 100 * np.mean(out_list)

    print(f"Validation Middlebury{split}: EPE {epe}, D1 {d1}")
    return {f'middlebury{split}-epe': epe, f'middlebury{split}-d1': d1}
是采用这样的方式，进行验证的吗，模型推理时的maxdisp相对于F\H\Q分辨率的参数都是192吗

**Comment by HalleyJiang** [comment]
我们是参考原RAFT-Stereo的评测代码，Middlebury 的所有分辨率都最大视差都截断都使用1000，并且也包括了遮挡像素。这个可以容纳了大视差和遮挡的因素，因此评测较为全面。这些论文里都zero-shot表格附近都写有，对于之前的方法我们也是按照这种方式在其开源的sceneflow 的checkpoint上评测。

**Comment by yyYcoco2** [comment]
感谢您的回答，我还是存在一些疑惑：
raft是这样的：val = (valid_gt.reshape(-1) >= -0.5) & (flow_gt[0].reshape(-1) > -1000)，看您的是val = (valid.reshape(-1) >= 0.5) & (disp_gt.reshape(-1) < 1000)   (valid.reshape(-1) >= 0.5) 是不是代表的是无遮挡呀

**Comment by HalleyJiang** [comment]
具体请运行一下代码，可视化一下的mask即可一目了然。(disp_gt.reshape(-1) < 1000)是保留GT视差小于1000的作为有效值，实际上GT视差大于1000的很少。有些工作只保留小于192的像素，但Middlebury本身是高分辨率和大视差的，这样不利于评价模型在大视差条件下的表现。 raftstereo那里用的是flow_gt相当于视差的负数，所以取反了。Middlebury提供了no_occ的一个mask，但我们都没用，所以评测时，把被遮挡的像素都包括进来了。
